src/python/pants/engine/isolated_process.py

```diff
@@ -19,18 +19,20 @@
 
 
 class ExecuteProcessRequest(datatype([
-  ('argv', tuple),
+  ('argv', TypedCollection(Exactly(text_type))),
   ('input_files', Digest),
   ('description', text_type),
-  ('env', tuple),
-  ('output_files', tuple),
-  ('output_directories', tuple),
+  ('env', TypedCollection(Exactly(text_type))),
+  ('output_files', TypedCollection(Exactly(text_type))),
+  ('output_directories', TypedCollection(Exactly(text_type))),
```

Comment:
Great! This is one of the things I'd most like to see expanded with our datatype system, to allow complex types like `Tuple[int, str]`, `Dict[str, int]`, `Optional[MyObject]`. Maybe something we can achieve by merging with MyPy and type hints.

Reply:
Ideally we just do something like #7074, and hope that dataclasses are something that easily extends to mypy, or we write a mypy plugin -- I'd prefer the first.